src/System.Private.ServiceModel/src/Internals/System/Runtime/TimeoutHelper.cs

```diff
@@ -1,12 +1,81 @@
 // Copyright (c) Microsoft. All rights reserved.
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
+using System.Collections.Concurrent;
 using System.Diagnostics.Contracts;
-using System;
 using System.Threading;
+using System.Threading.Tasks;
 
 namespace System.Runtime
 {
+    /// <summary>
+    /// This class coalesces timeout tokens because cancelation tokens with timeouts are more expensive to expose.
+    /// Disposing too many such tokens will cause thread contentions in high throughput scenario.
+    ///
+    /// Tokens with target cancelation time 15ms apart would resolve to the same instance.
+    /// </summary>
+    internal static class TimeoutTokenSource
```

Comment:
As a final step (otherwise it will make reviewing harder), can you move TimeoutHelper to be the first class in this file. It's a personal style thing, but I like to have the first class in a file match the filename.